landoapi/stacks.py

```diff
@@ -45,36 +42,13 @@ def build_stack_graph(
         between two nodes. `child` and `parent` are also string
         PHIDs.
     """
-    phids = set()
-    new_phids = {revision_phid}
-    edges = []
-
-    # Repeatedly request all related edges, adding connected revisions
-    # each time until no new revisions are found.
-    while new_phids:
-        phids.update(new_phids)
-        edges = phab.call_conduit(
-            "edge.search",
-            types=["revision.parent", "revision.child"],
-            sourcePHIDs=[phid for phid in phids],
-            limit=10000,
-        )
-        edges = phab.expect(edges, "data")
-        new_phids = set()
-        for edge in edges:
-            new_phids.add(edge["sourcePHID"])
-            new_phids.add(edge["destinationPHID"])
-
-        new_phids = new_phids - phids
-
-    # Treat the stack like a commit DAG, we only care about edges going
-    # from child to parent. This is enough to represent the graph.
-    edges = {
-        (edge["sourcePHID"], edge["destinationPHID"])
-        for edge in edges
-        if edge["edgeType"] == "revision.parent"
-    }
+    stack_graph = PhabricatorClient.expect(revision, "fields", "stackGraph")
+    phids = set(stack_graph.keys())
+    edges = set()
 
+    for source_node, destination_nodes in stack_graph.items():
+        for predecessor_node in destination_nodes:
+            edges.add((source_node, predecessor_node))
```

Comment:
Good point -- I think the main thing that could be simplified here is `destination_nodes` and we can drop the "source" reference. I think you are referring to the child/parent, ancestor/descendant terminology which we got rid of, and which implied a hierarchy that isn't implied by successor/predecessor.

Source/destination here were intended as more neutral terminology within the scope of the for loop. Maybe this alternative approach is easier to read (and illustrates what I'm trying to get at here).

```python
for node, predecessors in stack_graph.items():
    for predecessor in predecessors:
        edges.add((node, predecessor))
```

It wouldn't make sense to use both "successor" and "predecessor" reference to describe an edge, since those words describe a relative relationship -- and we lose the "relative to what" in that case.

Reply:
That looks good to me. :+1: